test/diameter_avp_filter_SUITE.erl

```diff
@@ -0,0 +1,403 @@
+%%% Helpers
+%%%===================================================================
+
+init_session(SessionMod) ->
+    Defaults = #{'3GPP-GGSN-Address' => {172, 20, 16, 28},
+		 '3GPP-IMEISV' => <<82, 21, 50, 96, 32, 80, 30, 0>>,
+		 '3GPP-IMSI' => <<"250071234567890">>,
+		 '3GPP-Charging-Id' => 3604013806,
+		 '3GPP-IMSI-MCC-MNC' => <<"25999">>,
+		 '3GPP-GGSN-MCC-MNC' => <<"25888">>,
+		 '3GPP-MS-TimeZone' => {128, 1},
+		 '3GPP-MSISDN' => <<"46702123456">>, '3GPP-NSAPI' => 5,
+		 '3GPP-PDP-Type' => 'IPv4', '3GPP-RAT-Type' => 6,
+		 '3GPP-SGSN-Address' => {192, 168, 1, 1},
+		 '3GPP-SGSN-MCC-MNC' => <<"26201">>,
+		 '3GPP-Selection-Mode' => 0,
+		 '3GPP-User-Location-Info' =>
+		     <<24, 98, 242, 16, 64, 163, 98, 242, 16, 1, 156, 232,
+		       0>>,
+		 'Called-Station-Id' => <<"some.station.gprs">>,
+		 'Calling-Station-Id' => <<"543148000012345">>,
+		 'Framed-IP-Address' => {10, 106, 14, 227},
+		 'Framed-Protocol' => 'GPRS-PDP-Context',
+		 'Multi-Session-Id' =>
+		     1012552258277823040188863251876666193415858290601,
+		 'Username' => <<"ergw">>, 'Password' => <<"ergw">>,
+		 'Service-Type' => 'Framed-User',
+		 'Node-Id' => <<"PGW-001">>,
+		 'PDP-Context-Type' => primary,
+		 'Charging-Rule-Base-Name' => <<"m2m0001">>,
+		 '3GPP-GPRS-Negotiated-QoS-Profile' =>
+		     <<11, 146, 31, 147, 150, 64, 64, 255, 255, 255, 255, 17,
+		       1, 1, 64, 64>>,
+		 '3GPP-Allocation-Retention-Priority' => 2,
+		 '3GPP-Charging-Characteristics' => <<8, 0>>,
+		 'QoS-Information' =>
+		     #{'QoS-Class-Identifier' => 8,
+		       'Max-Requested-Bandwidth-DL' => 0,
+		       'Max-Requested-Bandwidth-UL' => 0,
+		       'Guaranteed-Bitrate-DL' => 0,
+		       'Guaranteed-Bitrate-UL' => 0,
+		       'Allocation-Retention-Priority' =>
+			   #{'Priority-Level' => 10,
+			     'Pre-emption-Capability' => 1,
+			     'Pre-emption-Vulnerability' => 0},
+		       'APN-Aggregate-Max-Bitrate-DL' => 84000000,
+		       'APN-Aggregate-Max-Bitrate-UL' => 8640000}},
+    maps:merge(Defaults, SessionMod).
+
+test_server_request(#diameter_packet{msg = [_ | Msg]} =	Packet,
+		    Svc, {_, Caps} = Peer, Extra) ->
+    #{'Session-Id' := SId,
+      'Auth-Application-Id' := AppId,
+      'CC-Request-Type' := Type,
+      'CC-Request-Number' := Number} = Msg,
+
+    Resp = case get_diameter_session_handler(SId) of
+	     Fun when is_function(Fun) ->
+		 Fun(Packet, Svc, Peer, Extra);
+	     _ -> ok
+	   end,
+
+    case Resp of
+      ok ->
+	#diameter_caps{origin_host = {OH, _},
+		       origin_realm = {OR, _}} = Caps,
+	CCA = #{'Session-Id' => SId,
+		'Auth-Application-Id' => AppId,
+		'Origin-Host' => OH,
+		'Origin-Realm' => OR,
+		'CC-Request-Type' => Type,
+		'CC-Request-Number' => Number,
+		'Result-Code' => ?'DIAMETER_BASE_RESULT-CODE_SUCCESS'},
+	{reply, ['CCA' | CCA]};
+      _ -> Resp
+    end.
+
+do_filter_test(Filter) ->
+    do_filter_test(Filter, #{}).
+
+do_filter_test(Filter, SessionMod) ->
```

Comment:
Isn't this a bit too complex?

You are basically trying to test filter_avps/2, right? That could be accomplished much simpler if you where to export filter_avps/2 in the test profile and call it directly from the test suite.

e.g.:
```
-ifdef(TEST).
-export([filter_avps/2]).
-endif.
```

Reply:
Yes, completely right, if I were to test only the function indeed. I wanted to do it end to end to confirm/prove that the manipulation of the AVP structure does not have other unexpected effects in the handling of the message by the diameter application. However if you deem that unit testing the function itself is sufficient, I'll simplify it accordingly.